rusqlite_migration/src/loader.rs

```diff
@@ -0,0 +1,80 @@
+use std::collections::BTreeMap;
+
+use crate::{Error, Result, M};
+use include_dir::{Dir, DirEntry};
+use log::trace;
+
+struct MigrationFile {
+    up: &'static str,
+    down: Option<&'static str>,
+}
+
+pub fn from_dir(dir: &'static Dir<'static>) -> Result<Vec<M<'static>>> {
+    let mut migrations: BTreeMap<usize, MigrationFile> = BTreeMap::new();
```

Comment:
Oh yes! I haven't thought about using it as such. Any ideas how the interface of that might look like?

What I thought about just now is create a new version of `from_directory`:
```rust
fn from_directory_with_hooks(path: &str, hooks: &BTreeMap<usize, {struct with up/down MigrationHooks}>) {
// Creates the map for migration files, then merges the provided hooks into it based on the id
}
```

Reply:
@cljoly opinions?

Reply:
Sorry for the delay to come back to you on this.

# First Option

> What I thought about just now is create a new version of from_directory:
>
> ```rust
> fn from_directory_with_hooks(path: &str, hooks: &BTreeMap<usize, {struct with up/down MigrationHooks}>) {
> // Creates the map for migration files, then merges the provided hooks into it based on the id
> }
> ```

We would probably need more in `{struct with up/down MigrationHooks}` (let’s name it **`MEdit`**), like a boolean for `foreign_key_check()` but that’s fine

I guess it would be very manual for the caller, if they had to handle the BTreeMap and the subtleties (like handling duplicated files) of building it themselves. We can solve that with a function to build a basic BTreeMap that the caller would then pass on.

# Second Option

Initially, I had in mind something like this:

```rust
fn main() {
    let migrations: Migrations = MigrationsBuilder::from_directory(todo!())
        .edit(1, |m: M| -> M { m.add_down_hook(|| Ok(())) })
        .edit(2, |m| M.add_up_hook(|| Ok(())).foreign_key_check())
        .finalize();
}
```

So `MigrationsBuilder` would hold a `BTreeMap` internally and it would use [`and_modify`](https://doc.rust-lang.org/std/collections/btree_map/enum.Entry.html#method.and_modify) on entries to pass an owned value to the closures. `finalize` would then collect the `BTreeMap` into a vector.

The advantage is that it remains close to the “normal” version and we can reuse the same methods like `foreign_key_check`. The drawback is that we need to introduce methods like `add_up_hook` that are slightly different from `up_with_hook`, but we can mitigate this by putting them into a separate `impl` block, only available with the `from-directory` feature. And of course the documentation should mention that in general, `up_with_hook` is more appropriate.

I’m not sure the types and ownership would work though.

# Third Option

Maybe we can merge the two and pass a struct like `MEdit` to avoid the ownership issues? A bit like this:

```rust
fn main() {
    let migrations: Migrations = MigrationsBuilder::from_directory(todo!())
        .edit(1, MEdit {down_hook: || Ok(()), ..Default::default() })
        .edit(2, |m| MEdit { up_hook: || Ok(()), foreign_key_check: true, ..Default::default() })
        .finalize();
}
```
I don’t particularly like the added complexity of doing the same thing entirely differently though.
The second option is better in that regard.

Reply:
What do you think? There may be other things that I’m overlooking…

Reply:
For the type difficulty of the **Second Option**, we can solve this by having the builder hold two maps. One (let’s call it `files_map`) is built straight from the files and has all the `M`s build from the file, so only `up` and `down` effectively. The second (`edits_map`) starts as an empty map and is filled every time a modification is made.

So conceptually the `edit` method would look like this:
```rust
fn edit(mut self, id: usize, f: Fn(M) -> M) -> mut self {
  let entry = self.files_map.entry(id);
  // If entry is unpopulated, return an error, otherwise
  // remove the entry from files_map, that gives us an owned m (of type M)
  self.edits_map.insert(id, f(m)); // Handle the case of double inserts too
  self
}
```
and `finalize`:
```rust
fn finlize(self) -> Migrations {
  // create a vec and fill it with entries from the edits_map (if 
  // there is an edited M there) or insert the entry at the same
  // position from files_map. We can relay on (and document) the
  // assumption that files_map holds keys from 0..n, with all
  // the numbers between 0 and n.
}

On the second modification of a migration on the same id, an error would be returned.

Reply:
I like the `Builder` idea, seems like something that should replace `Migrations::new` in a potential V2 as the baseline for creating `Migrations`.

The second option seems like the more ergonomic of the two. I'll try to implement it.

Reply:
@cljoly should we completely remove the `Migrations::from_directory` method and move it to `MigrationsBuilder`?

Reply:
> I like the `Builder` idea, seems like something that should replace `Migrations::new` in a potential V2 as the baseline for creating `Migrations`.

I’m not sure, a `Builder` constructor is essentially slightly more complicated than a direct constructor and we should try to keep the API as simple as possible. Plus I don’t think that it is uncommon, with one simpler than the other. See for instance [`Reqwest::Client`](https://docs.rs/reqwest/0.11.13/reqwest/struct.Client.html#method.new)

Reply:
> should we completely remove the Migrations::from_directory method and move it to MigrationsBuilder?

Yes I think so. There is the problem of the async equivalent on `AsyncMigrations` though. Maybe the builder accommodate that by 1. existing across all features (even without methods for now) and 2. handling the creation of an `AsyncMigrations` struct as well, maybe with a different finalizer.

Reply:
> Also: What do you think about changing the Migrations::ms field to be a BTreeMap<NonZeroUsize, M<'m>>?

Is there any particular benefit you hope to get by doing that?

Reply:
> Is there any particular benefit you hope to get by doing that?

Not necessarily, I see some potential for code reuse, but it's rather minuscule.

Reply:
Ah ok. Yeah, we would avoid building a `Vec` here. We would still have to merge the two btrees though. And the “normal” case would be worse, because I think it’s often more costly to iterate over a BTree than it is to iterate over a `Vec`. Plus a BTree tends to occupy a bit more memory (due to the metadada of the tree itself) so it’s not great for the `Migrations`, that can potentially stay around for the duration of the whole program.

Reply:
I pushed a WIP for the time being, gonna have another look at it tomorrow. I store the migrations as a `Vec` so I can reuse the `from_directory` method in all cases and it will be slightly faster than a `BTreeMap` (wasn't an influence on my choice, but it is a benefit! :D)

I think I also found a solution for sync/async reusability by defining the Finalize method as:
```rust
pub fn finalize<T: FromIterator<M<'u>>>(mut self) -> T
```
and implementing `FromIterator` for both `Migrations` and `AsyncMigrations`. This allows us to just choose the result when the builder is used.

There's still some issue with lifetimes in the `edit` function, but I didn't have a deeper look into it yet.

Reply:
Sounds good!

Reply:
Thanks

Reply:
I looked into the edit lifetime issue.

So we need to get a owned M to edit it. We would need to remove it from the vector and leave a hole or a default value in its place, right? The problem is, I could not find a method for Vec that does just that. I see two option:

1. do something with [swap_remove](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.swap_remove) and [swap](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.swap) to filter the element through f with ownership and put the result in its place. That might be a bit brittle.
2. Use a btree. And then, instead of having another btree for edits as I was suggesting yesterday, you can actually just have one and edit by removing and inserting the modified version in the same place.

I would prefer 2 since it is easier to follow and to implement correctly (off by one errors…), but if you want to make a case for 1, we can try.

Reply:
(I have not looked into all of your latest changes, but hopefully the above will still help)

Reply:
No pressure @Czocher, you can do this in your own time, but I’ll do a deeper review once you have figured out the lifetime stuff in this thread.

Reply:
Sorry for the delay, unfortunately I'm having a rather busy end of the month. Thanks for being patient!

I'd like to experiment with something too before we merge this PR. I noticed other migration systems use a different directory structure - a set of numbered directories containing `up.sql` and `down.sql` files.
I'd like to check how this would look like, since the current implementation tickles me in a negative way in a few cases.

Reply:
> Sorry for the delay, unfortunately I'm having a rather busy end of the month

That’s ok, we all do that on our spare time and for fun, so take all the time you need.

> I'd like to experiment with something too before we merge this PR

Ok, cool! As long as we have not made a stable release though, we have total freedom to change anything in this PR, if we get a better idea.

> a set of numbered directories containing up.sql and down.sql files.
I'd like to check how this would look like, since the current implementation tickles me in a negative way in a few cases.

👍🏼 
Sure, I’m of course happy to consider other things too. I guess the potentially duplicated `up`/`down` issue would be more easily solved with that.

As a third option, another way to get that benefit would be to name the files `number-up/down.sql` like `02-up.sql` or `03-down.sql`.

Reply:
@cljoly hah good timing - I just started working on it again :D. 

I implemented the changes, seems it removed some edge-cases (just as you noticed) and simplified the code. I believe this was a good course of action.

Additionally I implemented the `MigrationsBuilder` (without tests at the moment). Can you have a look and share your comments?

Reply:
Thanks, I’ll take a look when I get the chance, mosty likely this week. I’ve been quite busy with the hotfix 🔥